meson: Use system doctest if available

nico202 opened the pull request.

nico202 committed:
meson: Use system doctest if available

* meson.build (doctest_proj): check if available as system dependency
* blurhash.cpp: change doctest include path to doctest/doctest.h

nico202 commented:
Wait, this does not seem to be enough as currently doctest is included with <doctest.h> while it should be doctest/doctest.h
like in https://github.com/onqtam/doctest/blob/a82821424d4826bfa891b4afef6767f447012774/examples/all_features/doctest_proxy.h#L6

nico202 commented:
I can now build it both with system version and with meson subproject

Commit:
meson: install pkgconfig and header files

nico202 commented:
It should also install pkgconfig and header files so that it can be found by other packages, as done in the second unrelated commit (I might split the pr in 2 separates if you prefer)

deepbluev7 commented on meson.build:
```diff
@@ -6,8 +6,13 @@ lib = static_library('blurhash',
 
-doctest_proj = subproject('doctest')
-doctest_dep = doctest_proj.get_variable('doctest_dep')
+doctest_dep = dependency('doctest', required : false)
+
+if not doctest_dep.found()
+	doctest_proj = subproject('doctest')
+	doctest_dep = doctest_proj.get_variable('doctest_dep')
+endif
+
```
```suggestion
doctest_dep = dependency('doctest', fallback : ['doctest', 'doctest_dep'])

```

Should do the same right? In theory we could also use a wrap with a provides section like here: https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section But that probably is too much to change atm, so the above is fine for now.

nico202 replied:
Did not know about fallback, tested both cases and it build just fine, applied and squashed

deepbluev7 reviewed (commented):
Looks good! Thank you for picking up my slack for me!

deepbluev7 commented on blurhash.cpp:
```diff
@@ -11,7 +11,7 @@
 
 #ifdef DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
-#include <doctest.h>
+#include <doctest/doctest.h>
```
This actually does not seem to work. The upstream meson file ( https://github.com/onqtam/doctest/blob/master/meson.build ), sets the include folder to doctest, which means you can't find the doctest.h, if you use the subproject. Are you sure it doesn't just work on your system, because it looks in /usr/include? I'm going to check what the installed version of doctest does now, if I can, but I think it sets the include path similarly. Otherwise we'll need to do some __has_include hackery.

deepbluev7 replied:
Okay, verifying that it works with a globally installed doctest will be a bit more effort for me, maybe we should do the following in the meantime?

```suggestion
#if __has_include(<doctest.h>)
#include <doctest.h>
#else
#include <doctest/doctest.h>
#endif
```

nico202 replied:
If it's fine for you to have this hack it's fine, else I'll patch this file in guix build so that you have more time to address this in the "right" way

deepbluev7 replied:
We have this hack for lmdbxx in Nheko already. Sadly for single header libs their uninstalled version often slightly differs from the installed version... Thank you!

nico202 committed:
Update blurhash.cpp

Apply suggestion from github interface (will need to be squashed before merge)

Co-authored-by: DeepBlueV7.X <nicolas.werner@hotmail.de>
